File: server/services/groqService.py
# -*- coding: utf-8 -*-
"""
NAZAR Backend Groq Conversational AI Service (Python Implementation)
Model: llama-3.1-8b-instant

Provides persistent quota tracking, automatic API key rotation at 14,000 requests,
daily UTC reset, and 429 failover without exposing API keys or crashing.
"""
import os
import json
import time
import urllib.request
import urllib.error
from datetime import datetime, timezone

SINGLETON_ID = "groq_usage"
ROTATION_LIMIT = 14000
LOCAL_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "../data/groq_usage.json")

def get_today_utc():
    return datetime.now(timezone.utc).strftime("%Y-%m-%d")

def discover_keys():
    keys = {}
    k1 = os.environ.get("GROQ_API_KEY_1", "").strip()
    k2 = os.environ.get("GROQ_API_KEY_2", "").strip()
    if k1: keys[1] = k1
    if k2: keys[2] = k2
    return keys

in_memory_state = {
    "activeKey": 1,
    "key1Requests": 0,
    "key2Requests": 0,
    "lastReset": get_today_utc()
}

def sync_state():
    today = get_today_utc()
    if in_memory_state["lastReset"] != today:
        print(f"[GroqService.py] Daily UTC midnight reset triggered. Date: {today}")
        in_memory_state["key1Requests"] = 0
        in_memory_state["key2Requests"] = 0
        in_memory_state["activeKey"] = 1
        in_memory_state["lastReset"] = today

    try:
        if os.path.exists(LOCAL_FILE):
            with open(LOCAL_FILE, "r", encoding="utf-8") as f:
                saved = json.load(f)
            if saved.get("lastReset") != today:
                saved["key1Requests"] = 0
                saved["key2Requests"] = 0
                saved["activeKey"] = 1
                saved["lastReset"] = today
            in_memory_state["activeKey"] = saved.get("activeKey", 1)
            in_memory_state["key1Requests"] = saved.get("key1Requests", 0)
            in_memory_state["key2Requests"] = saved.get("key2Requests", 0)
            in_memory_state["lastReset"] = saved.get("lastReset", today)
    except Exception:
        pass

def save_state(reason=""):
    try:
        dir_path = os.path.dirname(LOCAL_FILE)
        if not os.path.exists(dir_path):
            os.makedirs(dir_path, exist_ok=True)
        with open(LOCAL_FILE, "w", encoding="utf-8") as f:
            json.dump(in_memory_state, f, indent=2)
    except Exception as e:
        print("[GroqService.py] Failed to save state:", e)

def get_active_key_number():
    sync_state()
    k_num = in_memory_state["activeKey"]
    current_reqs = in_memory_state["key1Requests"] if k_num == 1 else in_memory_state["key2Requests"]
    if current_reqs >= ROTATION_LIMIT:
        print(f"[GroqService.py] Quota limit ({ROTATION_LIMIT}) reached on Key #{k_num}. Rotating...")
        k_num = 2 if k_num == 1 else 1
        in_memory_state["activeKey"] = k_num
        save_state(f"Quota reached on Key #{2 if k_num == 2 else 1}")
    return k_num

def record_success(key_num):
    if key_num == 1:
        in_memory_state["key1Requests"] += 1
    else:
        in_memory_state["key2Requests"] += 1
    save_state("Request success")
    print(f"[GroqService.py] Recorded success on Key #{key_num}. Total -> Key #1: {in_memory_state['key1Requests']}, Key #2: {in_memory_state['key2Requests']}")

def rotate_key(reason="Manual rotation"):
    old_key = in_memory_state["activeKey"]
    in_memory_state["activeKey"] = 2 if old_key == 1 else 1
    print(f"[GroqService.py] Rotating API Key: #{old_key} -> #{in_memory_state['activeKey']}. Reason: {reason}")
    save_state(reason)
    return in_memory_state["activeKey"]

def execute_call(key_num, messages, tools=None, tool_choice="auto", temperature=0.1):
    keys_map = discover_keys()
    api_key = keys_map.get(key_num) or keys_map.get(1) or keys_map.get(2)
    if not api_key:
        raise Exception(f"Missing API Key #{key_num}")
    
    model = os.environ.get("GROQ_MODEL", "llama-3.1-8b-instant")
    url = "https://api.groq.com/openai/v1/chat/completions"
    payload = {
        "model": model,
        "messages": messages,
        "temperature": temperature
    }
    if tools and len(tools) > 0:
        payload["tools"] = tools
        payload["tool_choice"] = tool_choice

    data_bytes = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(url, data=data_bytes, headers={
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    })
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        err = Exception(f"HTTP {e.code}: {e.read().decode('utf-8', errors='ignore')}")
        err.status = e.code
        raise err

def generate_response(options=None, **kwargs):
    if options is None:
        options = kwargs
    messages = options.get("messages", [])
    tools = options.get("tools", None)
    tool_choice = options.get("tool_choice", "auto")
    temperature = options.get("temperature", 0.1)

    keys_map = discover_keys()
    if not keys_map:
        print("[GroqService.py] No GROQ_API_KEY configured.")
        return {
            "success": False,
            "error": True,
            "message": "The assistant is temporarily busy. Please try again in a few minutes.",
            "friendlyResponse": True
        }

    current_key_num = get_active_key_number()
    try:
        data = execute_call(current_key_num, messages, tools, tool_choice, temperature)
        record_success(current_key_num)
        return {
            "success": True,
            "data": data,
            "keyUsed": current_key_num
        }
    except Exception as err:
        status = getattr(err, "status", None)
        print(f"[GroqService.py] Error on Key #{current_key_num}: HTTP {status or 'unknown'}")
        if status == 429 or (status and status >= 500) or not status:
            print(f"[GroqService.py] Triggering failover rotation due to error ({status or err})...")
            next_key_num = rotate_key(f"Failover from Key #{current_key_num} due to HTTP {status or 'error'}")
            try:
                print(f"[GroqService.py] Retrying request once on Key #{next_key_num}...")
                retry_data = execute_call(next_key_num, messages, tools, tool_choice, temperature)
                record_success(next_key_num)
                return {
                    "success": True,
                    "data": retry_data,
                    "keyUsed": next_key_num
                }
            except Exception as retry_err:
                r_status = getattr(retry_err, "status", None)
                print(f"[GroqService.py] Retry failed on Key #{next_key_num}: HTTP {r_status or 'unknown'}")
        
        return {
            "success": False,
            "error": True,
            "message": "The assistant is temporarily busy. Please try again in a few minutes.",
            "friendlyResponse": True
        }

def get_usage():
    return {
        "activeKey": in_memory_state["activeKey"],
        "key1Requests": in_memory_state["key1Requests"],
        "key2Requests": in_memory_state["key2Requests"],
        "lastReset": in_memory_state["lastReset"],
        "model": os.environ.get("GROQ_MODEL", "llama-3.1-8b-instant")
    }

File: server/services/groqService.js
/**
 * NAZAR Backend Groq Conversational AI Service
 * v1.0.0
 * 
 * Replaces legacy conversational AI backend with Groq's llama-3.1-8b-instant.
 * Provides automatic key rotation, persistent quota tracking (MongoDB + Local JSON fallback),
 * daily UTC reset, and 429 failover without exposing API keys or crashing.
 */
const fs = require('fs');
const path = require('path');
const ApiKeyUsage = require('../models/ApiKeyUsage');

const SINGLETON_ID = 'groq_usage';
const ROTATION_LIMIT = 14000; // Quota safety buffer below 14,400 daily limit
const LOCAL_FILE = path.join(__dirname, '../data/groq_usage.json');

// Get current date string in UTC (YYYY-MM-DD)
function getTodayUTC() {
    return new Date().toISOString().split('T')[0];
}

// Discover configured Groq API keys from environment variables
function discoverKeys() {
    const keys = new Map();
    const k1 = (process.env.GROQ_API_KEY_1 || '').trim();
    const k2 = (process.env.GROQ_API_KEY_2 || '').trim();
    if (k1) keys.set(1, k1);
    if (k2) keys.set(2, k2);
    return keys;
}

// In-memory atomic state
let inMemoryState = {
    activeKey: 1,
    key1Requests: 0,
    key2Requests: 0,
    lastReset: getTodayUTC()
};

// Cached HTTP client instances per API key to avoid re-creation per request
const clientCache = new Map();

class GroqClient {
    constructor(apiKey, model) {
        this.apiKey = apiKey;
        this.model = model;
        this.baseUrl = 'https://api.groq.com/openai/v1/chat/completions';
    }

    async chatCompletion(payload, timeoutMs = 15000) {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), timeoutMs);

        try {
            const res = await fetch(this.baseUrl, {
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${this.apiKey}`,
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    model: this.model,
                    ...payload
                }),
                signal: controller.signal
            });
            let data = {};
            try {
                data = await res.json();
            } catch (e) {
                data = {};
            }
            return { ok: res.ok, status: res.status, data };
        } finally {
            clearTimeout(timer);
        }
    }
}

function getClient(keyNum, apiKey) {
    const model = process.env.GROQ_MODEL || 'llama-3.1-8b-instant';
    const cacheKey = `${keyNum}:${model}`;
    if (!clientCache.has(cacheKey) || clientCache.get(cacheKey).apiKey !== apiKey) {
        clientCache.set(cacheKey, new GroqClient(apiKey, model));
    }
    return clientCache.get(cacheKey);
}

// Sync memory state with MongoDB and Local JSON file, handling daily UTC midnight reset
async function syncState() {
    const today = getTodayUTC();

    // Check memory date reset
    if (inMemoryState.lastReset !== today) {
        console.log(`[GroqService] Daily UTC midnight reset triggered. Date: ${today}`);
        inMemoryState.key1Requests = 0;
        inMemoryState.key2Requests = 0;
        inMemoryState.activeKey = 1;
        inMemoryState.lastReset = today;
    }

    try {
        let doc = await ApiKeyUsage.findOne({ singletonId: SINGLETON_ID });
        if (!doc) {
            doc = new ApiKeyUsage({
                singletonId: SINGLETON_ID,
                activeKey: inMemoryState.activeKey,
                activeModel: process.env.GROQ_MODEL || 'llama-3.1-8b-instant',
                keyUsage: { '1': inMemoryState.key1Requests, '2': inMemoryState.key2Requests },
                totalScans: 0,
                lastResetDate: today,
                rotationReason: 'Initial Groq setup'
            });
            await doc.save();
        } else if (doc.lastResetDate !== today) {
            doc.keyUsage.set('1', 0);
            doc.keyUsage.set('2', 0);
            doc.activeKey = 1;
            doc.lastResetDate = today;
            doc.rotationReason = 'Daily UTC quota reset';
            await doc.save();
        }

        inMemoryState.activeKey = doc.activeKey || 1;
        inMemoryState.key1Requests = doc.keyUsage.get('1') || 0;
        inMemoryState.key2Requests = doc.keyUsage.get('2') || 0;
        inMemoryState.lastReset = doc.lastResetDate || today;
    } catch (dbErr) {
        // Fallback to local JSON file if MongoDB is disconnected or unavailable
        try {
            if (fs.existsSync(LOCAL_FILE)) {
                const raw = fs.readFileSync(LOCAL_FILE, 'utf8');
                const saved = JSON.parse(raw);
                if (saved.lastReset !== today) {
                    saved.key1Requests = 0;
                    saved.key2Requests = 0;
                    saved.activeKey = 1;
                    saved.lastReset = today;
                }
                inMemoryState.activeKey = saved.activeKey || 1;
                inMemoryState.key1Requests = saved.key1Requests || 0;
                inMemoryState.key2Requests = saved.key2Requests || 0;
                inMemoryState.lastReset = saved.lastReset || today;
            }
        } catch (fileErr) {
            // Use existing inMemoryState
        }
    }
}

// Persist state to both local JSON file and MongoDB
async function saveState(reason = '') {
    const dir = path.dirname(LOCAL_FILE);
    try {
        if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
        fs.writeFileSync(LOCAL_FILE, JSON.stringify(inMemoryState, null, 2));
    } catch (e) {
        console.warn('[GroqService] Failed to write local JSON fallback file.');
    }

    try {
        let doc = await ApiKeyUsage.findOne({ singletonId: SINGLETON_ID });
        if (doc) {
            doc.activeKey = inMemoryState.activeKey;
            doc.keyUsage.set('1', inMemoryState.key1Requests);
            doc.keyUsage.set('2', inMemoryState.key2Requests);
            doc.lastResetDate = inMemoryState.lastReset;
            if (reason) {
                doc.lastRotation = new Date();
                doc.rotationReason = reason;
                doc.history.push({
                    from: doc.activeKey === 1 ? 2 : 1,
                    to: doc.activeKey,
                    reason: reason,
                    time: new Date()
                });
            }
            await doc.save();
        }
    } catch (e) {
        // MongoDB offline fallback handled by local file persistence
    }
}

// Check quota and rotate automatically at 14,000 requests
async function getActiveKeyNumber() {
    await syncState();
    let kNum = inMemoryState.activeKey;
    let currentReqs = kNum === 1 ? inMemoryState.key1Requests : inMemoryState.key2Requests;

    if (currentReqs >= ROTATION_LIMIT) {
        console.log(`[GroqService] Quota limit (${ROTATION_LIMIT}) reached on API Key #${kNum}. Automatically rotating to next key...`);
        kNum = kNum === 1 ? 2 : 1;
        inMemoryState.activeKey = kNum;
        await saveState(`Quota reached on Key #${kNum === 1 ? 2 : 1}`);
    }
    return kNum;
}

// Record successful request increment
async function recordSuccess(keyNum) {
    if (keyNum === 1) {
        inMemoryState.key1Requests++;
    } else {
        inMemoryState.key2Requests++;
    }
    await saveState('Request success');
    console.log(`[GroqService] Recorded success on Key #${keyNum}. Total Requests -> Key #1: ${inMemoryState.key1Requests}, Key #2: ${inMemoryState.key2Requests}`);
}

// Force rotate between configured keys
async function rotateKey(reason = 'Manual rotation') {
    const oldKey = inMemoryState.activeKey;
    inMemoryState.activeKey = oldKey === 1 ? 2 : 1;
    console.log(`[GroqService] Rotating API Key: #${oldKey} -> #${inMemoryState.activeKey}. Reason: ${reason}`);
    await saveState(reason);
    return inMemoryState.activeKey;
}

/**
 * Generate AI response using Groq llama-3.1-8b-instant with automatic key rotation and 429 failover.
 * @param {Object} options Options containing messages, tools, tool_choice, temperature
 * @returns {Promise<Object>} Result object with success status and completion data
 */
async function generate_response(options) {
    const { messages, tools, tool_choice = 'auto', temperature = 0.1 } = options;
    const keysMap = discoverKeys();

    if (keysMap.size === 0) {
        console.error('[GroqService] No GROQ_API_KEY configured in environment variables.');
        return {
            success: false,
            error: true,
            message: "The assistant is temporarily busy. Please try again in a few minutes.",
            friendlyResponse: true
        };
    }

    const executeCall = async (keyNum) => {
        const apiKey = keysMap.get(keyNum) || keysMap.get(1) || keysMap.get(2);
        if (!apiKey) {
            const err = new Error(`Missing API Key #${keyNum}`);
            err.status = 500;
            throw err;
        }
        const client = getClient(keyNum, apiKey);
        const res = await client.chatCompletion({
            messages,
            tools: tools && tools.length > 0 ? tools : undefined,
            tool_choice: tools && tools.length > 0 ? tool_choice : undefined,
            temperature
        });

        if (!res.ok) {
            const err = new Error(`HTTP ${res.status}: ${JSON.stringify(res.data)}`);
            err.status = res.status;
            err.data = res.data;
            throw err;
        }
        return res.data;
    };

    let currentKeyNum = await getActiveKeyNumber();

    try {
        const data = await executeCall(currentKeyNum);
        await recordSuccess(currentKeyNum);
        return {
            success: true,
            data,
            keyUsed: currentKeyNum
        };
    } catch (err) {
        console.warn(`[GroqService] Error on Key #${currentKeyNum}: HTTP ${err.status || 'unknown'}`);

        // Immediate failover on 429 Too Many Requests or server error
        if (err.status === 429 || err.status >= 500 || err.name === 'AbortError' || !err.status) {
            console.log(`[GroqService] Triggering failover rotation due to error (${err.status || err.message})...`);
            const nextKeyNum = await rotateKey(`Failover from Key #${currentKeyNum} due to HTTP ${err.status || 'error'}`);

            try {
                console.log(`[GroqService] Retrying request once on Key #${nextKeyNum}...`);
                const retryData = await executeCall(nextKeyNum);
                await recordSuccess(nextKeyNum);
                return {
                    success: true,
                    data: retryData,
                    keyUsed: nextKeyNum
                };
            } catch (retryErr) {
                console.error(`[GroqService] Retry failed on Key #${nextKeyNum}: HTTP ${retryErr.status || 'unknown'}`);
            }
        }

        // Both keys failed or unavailable: return friendly message without crashing
        return {
            success: false,
            error: true,
            message: "The assistant is temporarily busy. Please try again in a few minutes.",
            friendlyResponse: true
        };
    }
}

// Get current tracking analytics
function getUsage() {
    return {
        activeKey: inMemoryState.activeKey,
        key1Requests: inMemoryState.key1Requests,
        key2Requests: inMemoryState.key2Requests,
        lastReset: inMemoryState.lastReset,
        model: process.env.GROQ_MODEL || 'llama-3.1-8b-instant'
    };
}

module.exports = {
    generate_response,
    generateResponse: generate_response, // camelCase alias
    rotateKey,
    rotate_key: rotateKey,
    getUsage,
    get_usage: getUsage,
    discoverKeys
};
